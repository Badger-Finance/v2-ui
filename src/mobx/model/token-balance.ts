import BigNumber from 'bignumber.js';
import RewardsStore from 'mobx/stores/rewardsStore';
import { inCurrency, minBalance } from 'mobx/utils/helpers';
import { BadgerToken } from './badger-token';

export class TokenBalance {
	private store: RewardsStore;
	readonly token: BadgerToken;
	public tokenBalance: BigNumber;
	public balance: BigNumber;
	public price: BigNumber;

	constructor(store: RewardsStore, token: BadgerToken, balance: BigNumber, price: BigNumber) {
		this.store = store;
		this.token = token;
		this.tokenBalance = balance;
		this.price = price;
		this.balance = balance.dividedBy(Math.pow(10, token.decimals));
	}

	/**
	 * Create a token balance from a string balance of appropriate decimals.
	 * i.e.
	 *   Given token A as a defined TokenBalance.
	 *   TokenBalance.fromBalance(A, A.balanceDisplay()) = A;
	 * Above does not hold true given a balance display of balance below a
	 * a requested precision threshold (< 0.001 balance display).
	 * Does not work with digg share conversion - only fragments support.
	 */
	static fromBalance(tokenBalance: TokenBalance, balance: string): TokenBalance {
		const { token, store, price } = tokenBalance;
		const scalar = new BigNumber(Math.pow(10, token.decimals));
		const amount = new BigNumber(balance).multipliedBy(scalar);
		return new TokenBalance(store, token, amount, price);
	}

	get value(): BigNumber {
		return this.balance.multipliedBy(this.price);
	}

	/**
	 * Return a string display value for a given precision. Defaults to token decimals.
	 * i.e.
	 *  0 => 0.0 (1 decimal)
	 *  0.0001 => < 0.001 (3 decimals)
	 *  0.0001 => 0.00010 (5 decimals)
	 * @param precision decimal count for display purposes
	 * @returns string representation of balance
	 */
	balanceDisplay(precision?: number): string {
		const decimals = precision || this.token.decimals;
		if (this.balance.gt(0) && this.balance.lt(minBalance(decimals))) {
			return `< 0.${'0'.repeat(decimals - 1)}1`;
		}
		return this.balance.toFixed(decimals);
	}

	balanceValueDisplay(currency: string): string {
		return inCurrency(this.value, currency);
	}

	scale(scalar: BigNumber): TokenBalance {
		const tokenBalance = this.tokenBalance.multipliedBy(scalar);
		return new TokenBalance(this.store, this.token, tokenBalance, this.price);
	}

	scaledBalanceDisplay(percent: number): string {
		const scaledBalance = this.scale(new BigNumber(percent / 100));
		return scaledBalance.balanceDisplay();
	}
}
