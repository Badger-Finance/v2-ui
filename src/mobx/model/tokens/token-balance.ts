import { Token } from '@badger-dao/sdk';
import BigNumber from 'bignumber.js';
import { Currency } from 'config/enums/currency.enum';
import { inCurrency, minBalance } from 'mobx/utils/helpers';

export class TokenBalance {
	readonly token: Token;
	public tokenBalance: BigNumber;
	public balance: BigNumber;
	public price: BigNumber;

	constructor(token: Token, balance: BigNumber, price: BigNumber) {
		this.token = token;
		this.tokenBalance = balance;
		this.price = price;
		this.balance = balance.dividedBy(Math.pow(10, token.decimals));
	}

	/**
	 * Create a token balance from a string balance of appropriate decimals.
	 * i.e.
	 *   Given token A as a defined TokenBalance.
	 *   TokenBalance.fromBalance(A, A.balanceDisplay()) = A;
	 * Above does not hold true given a balance display of balance below a
	 * a requested precision threshold (< 0.001 balance display).
	 * Does not work with digg share conversion - only fragments support.
	 */
	static fromBalance(tokenBalance: TokenBalance, balance: string): TokenBalance {
		const { token, price } = tokenBalance;
		const scalar = new BigNumber(Math.pow(10, token.decimals));
		const amount = new BigNumber(balance).multipliedBy(scalar);
		return new TokenBalance(token, amount, price);
	}

	static fromBigNumber(tokenBalance: TokenBalance, balance: BigNumber): TokenBalance {
		const { token, price } = tokenBalance;
		return new TokenBalance(token, balance, price);
	}

	static hasBalance(tokenBalance?: TokenBalance): boolean {
		if (!tokenBalance) {
			return false;
		}
		return tokenBalance.balance.gt(0);
	}

	get value(): BigNumber {
		return this.balance.multipliedBy(this.price);
	}

	/**
	 * Return a string display value for a given precision. Defaults to token decimals.
	 * i.e.
	 *  0 => 0.0 (1 decimal)
	 *  0.0001 => < 0.001 (3 decimals)
	 *  0.0001 => 0.00010 (5 decimals)
	 * @param precision decimal count for display purposes
	 * @returns string representation of balance
	 */
	balanceDisplay(precision?: number): string {
		const decimals = precision || this.token.decimals;
		if (this.balance.gt(0) && this.balance.lt(minBalance(decimals))) {
			return `< 0.${'0'.repeat(decimals - 1)}1`;
		}

		return this.balance.toFixed(decimals);
	}

	balanceValueDisplay(currency: Currency): string | undefined {
		return inCurrency(this.value, currency);
	}

	/**
	 * Scale token balance by a given scalar.
	 * Conversions to bTokens can utilize this function with scalePrice
	 * to effectively convert balances using ppfs as the given scalar.
	 */
	scale(scalar: BigNumber, scalePrice?: boolean): TokenBalance {
		if (this.tokenBalance.eq(0)) {
			return this;
		}
		const tokenBalance = this.tokenBalance.multipliedBy(scalar);
		const price = scalePrice ? this.price.dividedBy(scalar) : this.price;
		return new TokenBalance(this.token, tokenBalance, price);
	}

	scaledBalanceDisplay(percent: number): string {
		const scaledBalance = this.scale(new BigNumber(percent / 100));
		return scaledBalance.balanceDisplay();
	}
}
