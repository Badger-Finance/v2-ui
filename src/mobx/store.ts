import { RouterStore } from 'mobx-router';
import UiState from './reducers';
import WalletStore from './stores/walletStore';
import ContractsStore from './stores/contractsStore';
import AirdropStore from './stores/airdropStore';
import RebaseStore from './stores/rebaseStore';
import RewardsStore from './stores/rewardsStore';
import IbBTCStore from './stores/ibBTCStore';
import TransactionsStore from './stores/transactionsStore';
import SettStoreV2 from './stores/settStoreV2';
import { NETWORK_LIST } from '../config/constants';
import HoneyPotStore from './stores/honeyPotStore';

export class RootStore {
	public router: RouterStore<RootStore>;
	public wallet: WalletStore;
	public uiState: UiState;
	public contracts: ContractsStore;
	public airdrops: AirdropStore;
	public rebase: RebaseStore;
	public rewards: RewardsStore;
	public ibBTCStore: IbBTCStore;
	public setts: SettStoreV2;
	public transactions: TransactionsStore;
	public honeyPot: HoneyPotStore;

	constructor() {
		this.router = new RouterStore<RootStore>(this);
		this.wallet = new WalletStore(this);
		this.contracts = new ContractsStore(this);
		this.airdrops = new AirdropStore(this);
		this.rebase = new RebaseStore(this);
		this.rewards = new RewardsStore(this);
		this.uiState = new UiState(this);
		this.ibBTCStore = new IbBTCStore(this);
		this.honeyPot = new HoneyPotStore(this);
		// RenVM transactions store.
		this.transactions = new TransactionsStore(this);
		this.setts = new SettStoreV2(this);

		this.walletRefresh();
	}

	async walletRefresh(): Promise<void> {
		const chain = this.wallet.network.name;
		const refreshData = [
			this.setts.loadAssets(chain),
			this.setts.loadPrices(chain),
			this.wallet.getGasPrice(),
			this.contracts.updateProvider(),
		];
		if (chain === NETWORK_LIST.ETH) {
			refreshData.push(this.setts.loadGeysers(chain));
			refreshData.push(this.rebase.fetchRebaseStats());
		} else {
			refreshData.push(this.setts.loadSetts(chain));
		}
		await Promise.all(refreshData);

		if (this.wallet.connectedAddress) {
			this.contracts.updateProvider();
			await this.wallet.getGasPrice();
			await this.contracts.fetchContracts();
			if (chain === NETWORK_LIST.ETH) {
				this.uiState.reduceRebase();
				this.ibBTCStore.init();
				this.rewards.fetchSettRewards();
				this.uiState.reduceTreeRewards();
				this.airdrops.fetchAirdrops();
				this.uiState.reduceAirdrops();
			}
			this.uiState.reduceStats();
		}
	}
}

const store = new RootStore();

export default store;
