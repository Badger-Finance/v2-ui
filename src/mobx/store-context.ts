import { RootStore } from './store';
import { createContext } from 'react';

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;
