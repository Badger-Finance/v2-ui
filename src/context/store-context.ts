import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;