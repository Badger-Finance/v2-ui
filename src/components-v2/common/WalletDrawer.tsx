import React, { useContext, useState } from 'react';
import { observer } from 'mobx-react-lite';
import { StoreContext } from '../../mobx/store-context';
import { Box, Button, Drawer, Grid, IconButton, makeStyles, Typography, useTheme } from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';
import useENS from '../../hooks/useEns';
import { shortenAddress } from '../../utils/componentHelpers';
import CurrencyDisplay from './CurrencyDisplay';
import { inCurrency } from '../../mobx/utils/helpers';
import copy from 'copy-to-clipboard';
import BigNumber from 'bignumber.js';

const useStyles = makeStyles((theme) => ({
	root: {
		width: 349,
		backgroundColor: theme.palette.background.paper,
		padding: '25px 25px 50px 25px',
		height: '100vh',
		position: 'relative',
	},
	title: {
		fontSize: 20,
		fontWeight: 700,
	},
	balance: {
		fontSize: 20,
	},
	disconnectWalletText: {
		color: '#FB0505',
		fontWeight: 700,
	},
	titleRow: {
		marginBottom: theme.spacing(2),
	},
	addressRow: {
		marginBottom: theme.spacing(2),
	},
	tokenNameAndIcon: {
		display: 'flex',
		alignItems: 'center',
	},
	tokenName: {
		fontSize: 16,
	},
	icon: {
		width: 25,
		marginRight: 15,
	},
	tokenBalance: {
		marginBottom: theme.spacing(3),
	},
	balancesList: {
		marginTop: 32,
	},
	balanceDisplayValue: {
		marginTop: theme.spacing(-0.5),
	},
	copiedMessage: {
		background: '#66BB6A',
		position: 'absolute',
		bottom: 0,
		paddingLeft: theme.spacing(2),
		paddingRight: theme.spacing(1),
	},
	closeButton: {
		position: 'absolute',
		top: theme.spacing(2),
		right: theme.spacing(1),
	},
	copyWalletButton: {
		marginLeft: theme.spacing(4),
	},
	disconnectWalletIcon: {
		marginRight: theme.spacing(3),
	},
}));

const WalletDrawer = (): JSX.Element | null => {
	const [showCopiedMessage, setShowCopiedMessage] = useState(false);
	const { uiState, user, onboard, network } = useContext(StoreContext);
	const { ensName } = useENS(onboard.address);
	const classes = useStyles();
	const closeDialogTransitionDuration = useTheme().transitions.duration.leavingScreen;

	const handleCopy = () => {
		if (!onboard.address) return;
		const didCopy = copy(onboard.address);
		setShowCopiedMessage(didCopy);
	};

	const handleDisconnect = () => {
		uiState.toggleWalletDrawer();
		setTimeout(() => {
			onboard.disconnect();
		}, closeDialogTransitionDuration);
	};

	if (!onboard.address) {
		return null;
	}

	const tokenBalances = Object.keys(network.network.deploy.tokens).flatMap((token) => {
		const isBadgerToken = ['badger', 'digg', 'remdigg'].includes(token.toLowerCase());
		return isBadgerToken ? [user.getTokenBalance(network.network.deploy.tokens[token])] : [];
	});

	const sortedBalances = tokenBalances.sort((a, b) => b.value.minus(a.value).toNumber());
	const totalBalance = tokenBalances.reduce((total, next) => total.plus(next.value), new BigNumber(0));

	return (
		<Drawer open={uiState.showWalletDrawer} anchor="right" onClose={() => uiState.toggleWalletDrawer()}>
			<Grid container direction="column" className={classes.root} justifyContent="space-between">
				<Grid item>
					<Grid
						item
						container
						justifyContent="space-between"
						alignItems="center"
						className={classes.titleRow}
					>
						<Typography variant="h6" display="inline" className={classes.title}>
							Wallet
						</Typography>
						<IconButton className={classes.closeButton} onClick={() => uiState.toggleWalletDrawer()}>
							<CloseIcon />
						</IconButton>
					</Grid>
					<Grid item container alignItems="center" className={classes.addressRow}>
						<Typography variant="subtitle2" display="inline">
							{ensName || shortenAddress(onboard.address)}
						</Typography>
						<IconButton
							onClick={handleCopy}
							aria-label="copy wallet address"
							className={classes.copyWalletButton}
						>
							<img src="/assets/icons/copy-wallet-address.svg" alt="copy wallet address icon" />
						</IconButton>
					</Grid>
					<Grid item>
						<Typography variant="subtitle2" color="textSecondary">
							Balance:
						</Typography>
						<CurrencyDisplay
							variant="body1"
							justifyContent="flex-start"
							displayValue={inCurrency(totalBalance, uiState.currency)}
							TypographyProps={{ className: classes.balance }}
						/>
					</Grid>
					<Grid item container className={classes.balancesList}>
						{sortedBalances.map((tokenBalance) => (
							<Grid container className={classes.tokenBalance} key={tokenBalance.token.address}>
								<Grid item container justifyContent="space-between" alignItems="center">
									<div className={classes.tokenNameAndIcon}>
										<img
											className={classes.icon}
											// small hack to allow rem tokens to share icons with their badger counterparts
											src={`/assets/icons/${tokenBalance.token.symbol
												.replace('rem', '')
												.toLowerCase()
												.trim()}.png`}
											alt={`${tokenBalance.token.name} icon`}
										/>
										<Typography variant="body1" display="inline" className={classes.tokenName}>
											{tokenBalance.token.symbol}
										</Typography>
									</div>
									<Box display="inline">
										<Typography variant="body1">{tokenBalance.balanceDisplay(6)}</Typography>
									</Box>
								</Grid>
								<Grid item container justifyContent="flex-end" className={classes.balanceDisplayValue}>
									<Typography variant="subtitle2" color="textSecondary">
										{tokenBalance.balanceValueDisplay(uiState.currency, 6)}
									</Typography>
								</Grid>
							</Grid>
						))}
					</Grid>
				</Grid>
				<Grid item>
					<Button
						variant="text"
						classes={{ label: classes.disconnectWalletText }}
						onClick={handleDisconnect}
						aria-label="disconnect wallet"
					>
						<img
							className={classes.disconnectWalletIcon}
							src="/assets/icons/disconnect-wallet.svg"
							aria-label="disconnect wallet"
							alt="disconnect wallet icon"
						/>
						Disconnect Wallet
					</Button>
				</Grid>
			</Grid>
			{showCopiedMessage && (
				<Grid container className={classes.copiedMessage} alignItems="center" justifyContent="space-between">
					<Typography variant="subtitle2" display="inline">
						Wallet Address Copied
					</Typography>
					<IconButton onClick={() => setShowCopiedMessage(false)} aria-label="dismiss copied address message">
						<CloseIcon />
					</IconButton>
				</Grid>
			)}
		</Drawer>
	);
};

export default observer(WalletDrawer);
