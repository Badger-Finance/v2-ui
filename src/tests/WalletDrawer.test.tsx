import React from 'react';
import '@testing-library/jest-dom';
import { OnboardStore } from '../mobx/stores/OnboardStore';
import store from '../mobx/RootStore';
import { StoreProvider } from '../mobx/store-context';
import WalletDrawer from '../components-v2/common/WalletDrawer';
import { customRender, screen, fireEvent } from './Utils';
import { TokenBalance } from '../mobx/model/tokens/token-balance';
import * as copy from 'copy-to-clipboard';
import UserStore from '../mobx/stores/UserStore';
import deploy from '../config/deployments/mainnet.json';
import BigNumber from 'bignumber.js';

jest.mock('copy-to-clipboard', () => {
	return jest.fn().mockReturnValue(true);
});

describe('Wallet Drawer', () => {
	beforeEach(() => {
		store.uiState.showWalletDrawer = true;
		store.onboard.address = '0x1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a';
		jest.spyOn(OnboardStore.prototype, 'isActive').mockReturnValue(true);
		jest.spyOn(UserStore.prototype, 'getTokenBalance').mockImplementation((token) => {
			if (token === deploy.tokens.badger) {
				return new TokenBalance(
					{
						address: '0x1',
						name: 'Badger',
						symbol: 'Badger',
						decimals: 18,
					},
					new BigNumber(1000 * 1e18),
					new BigNumber(80),
				);
			} else {
				return new TokenBalance(
					{
						address: '0x2',
						name: 'Digg',
						symbol: 'DIGG',
						decimals: 8,
					},
					new BigNumber(0.1 * 1e8),
					new BigNumber(50000),
				);
			}
		});
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	it('displays correctly', () => {
		const { baseElement } = customRender(
			<StoreProvider value={store}>
				<WalletDrawer />
			</StoreProvider>,
		);
		expect(baseElement).toMatchSnapshot();
	});

	it('disconnects wallet', () => {
		jest.useFakeTimers();
		const disconnectSpy = jest.fn();
		store.onboard.disconnect = disconnectSpy;

		customRender(
			<StoreProvider value={store}>
				<WalletDrawer />
			</StoreProvider>,
		);

		fireEvent.click(screen.getByRole('button', { name: 'disconnect wallet' }));
		jest.runAllTimers();
		expect(disconnectSpy).toHaveBeenCalled();
	});

	it('copies wallet address', () => {
		jest.spyOn(copy, 'default').mockReturnValue(true);

		customRender(
			<StoreProvider value={store}>
				<WalletDrawer />
			</StoreProvider>,
		);

		fireEvent.click(screen.getByRole('button', { name: 'copy wallet address' }));
		expect(screen.getByText('Wallet Address Copied')).toBeInTheDocument();
	});

	it('dismisses copied wallet address message', () => {
		jest.spyOn(copy, 'default').mockReturnValue(true);

		customRender(
			<StoreProvider value={store}>
				<WalletDrawer />
			</StoreProvider>,
		);

		fireEvent.click(screen.getByRole('button', { name: 'copy wallet address' }));
		fireEvent.click(screen.getByRole('button', { name: 'dismiss copied address message' }));
		expect(screen.queryByText('Wallet Address Copied')).not.toBeInTheDocument();
	});
});
