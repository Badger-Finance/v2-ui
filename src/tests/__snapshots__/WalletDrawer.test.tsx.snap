// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet Drawer displays correctly 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
   
   
  
  
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDrawer-root-15 MuiDrawer-modal-26"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root-27"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root-29 MuiDrawer-paper-17 MuiDrawer-paperAnchorRight-19 MuiPaper-elevation16-48"
      tabindex="-1"
    >
      <div
        class="MuiGrid-root-57 makeStyles-root-1 MuiGrid-container-58 MuiGrid-direction-xs-column-61 MuiGrid-justify-content-xs-space-between-77"
      >
        <div
          class="MuiGrid-root-57 MuiGrid-item-59"
        >
          <div
            class="MuiGrid-root-57 makeStyles-titleRow-6 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-align-items-xs-center-66 MuiGrid-justify-content-xs-space-between-77"
          >
            <h6
              class="MuiTypography-root-160 makeStyles-title-2 MuiTypography-h6-170 MuiTypography-displayInline-188"
            >
              Wallet
            </h6>
            <button
              class="MuiButtonBase-root-199 MuiIconButton-root-190 makeStyles-closeButton-11"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-198"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root-202"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root-266"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root-57 makeStyles-addressRow-7 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-align-items-xs-center-66"
          >
            <h6
              class="MuiTypography-root-160 makeStyles-address-14 MuiTypography-subtitle2-172 MuiTypography-colorTextSecondary-186 MuiTypography-displayInline-188"
            >
              0x1..a1a
            </h6>
            <button
              aria-label="copy wallet address"
              class="MuiButtonBase-root-199 MuiIconButton-root-190 makeStyles-copyWalletButton-12"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-198"
              >
                <img
                  alt="copy wallet address icon"
                  src="/assets/icons/copy-wallet-address.svg"
                />
              </span>
              <span
                class="MuiTouchRipple-root-266"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root-57 MuiGrid-item-59"
          >
            <h6
              class="MuiTypography-root-160 makeStyles-balanceText-3 MuiTypography-subtitle2-172 MuiTypography-colorTextSecondary-186"
            >
              Balance:
            </h6>
            <div
              class="MuiBox-root-213 MuiBox-root-214"
            >
              <p
                class="MuiTypography-root-160 makeStyles-balance-4 MuiTypography-body1-162"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root-57 makeStyles-balancesList-8 MuiGrid-container-58 MuiGrid-item-59"
          >
            <div
              class="MuiGrid-root-57 makeStyles-tokenBalance-220 MuiGrid-container-58"
            >
              <div
                class="MuiGrid-root-57 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-align-items-xs-center-66 MuiGrid-justify-content-xs-space-between-77"
              >
                <div
                  class="makeStyles-tokenNameAndIcon-216"
                >
                  <img
                    alt="Badger logo"
                    class="makeStyles-icon-218"
                    src="/assets/icons/badger.svg"
                  />
                  <p
                    class="MuiTypography-root-160 makeStyles-tokenName-217 MuiTypography-body1-162 MuiTypography-displayInline-188"
                  >
                    Badger
                  </p>
                </div>
                <div
                  class="MuiBox-root-213 MuiBox-root-223"
                >
                  <p
                    class="MuiTypography-root-160 makeStyles-balance-219 MuiTypography-body1-162"
                  >
                    1000.000000
                  </p>
                </div>
              </div>
              <div
                class="MuiGrid-root-57 makeStyles-balanceDisplayValue-222 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-justify-content-xs-flex-end-76"
              >
                <h6
                  class="MuiTypography-root-160 makeStyles-balance-219 MuiTypography-subtitle2-172 MuiTypography-colorTextSecondary-186"
                />
              </div>
            </div>
            <div
              class="MuiGrid-root-57 makeStyles-tokenBalance-220 MuiGrid-container-58"
            >
              <div
                class="MuiGrid-root-57 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-align-items-xs-center-66 MuiGrid-justify-content-xs-space-between-77"
              >
                <div
                  class="makeStyles-tokenNameAndIcon-216"
                >
                  <img
                    alt="DIGG logo"
                    class="makeStyles-icon-218"
                    src="/assets/icons/digg.svg"
                  />
                  <p
                    class="MuiTypography-root-160 makeStyles-tokenName-217 MuiTypography-body1-162 MuiTypography-displayInline-188"
                  >
                    DIGG
                  </p>
                </div>
                <div
                  class="MuiBox-root-213 MuiBox-root-224"
                >
                  <p
                    class="MuiTypography-root-160 makeStyles-balance-219 MuiTypography-body1-162"
                  >
                    0.100000
                  </p>
                </div>
              </div>
              <div
                class="MuiGrid-root-57 makeStyles-balanceDisplayValue-222 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-justify-content-xs-flex-end-76"
              >
                <h6
                  class="MuiTypography-root-160 makeStyles-balance-219 MuiTypography-subtitle2-172 MuiTypography-colorTextSecondary-186"
                />
              </div>
            </div>
            <div
              class="MuiGrid-root-57 makeStyles-tokenBalance-220 MuiGrid-container-58"
            >
              <div
                class="MuiGrid-root-57 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-align-items-xs-center-66 MuiGrid-justify-content-xs-space-between-77"
              >
                <div
                  class="makeStyles-tokenNameAndIcon-216"
                >
                  <img
                    alt="DIGG logo"
                    class="makeStyles-icon-218"
                    src="/assets/icons/digg.svg"
                  />
                  <p
                    class="MuiTypography-root-160 makeStyles-tokenName-217 MuiTypography-body1-162 MuiTypography-displayInline-188"
                  >
                    remDIGG
                  </p>
                </div>
                <div
                  class="MuiBox-root-213 MuiBox-root-225"
                >
                  <p
                    class="MuiTypography-root-160 makeStyles-balance-219 MuiTypography-body1-162"
                  >
                    0.100000
                  </p>
                </div>
              </div>
              <div
                class="MuiGrid-root-57 makeStyles-balanceDisplayValue-222 MuiGrid-container-58 MuiGrid-item-59 MuiGrid-justify-content-xs-flex-end-76"
              >
                <h6
                  class="MuiTypography-root-160 makeStyles-balance-219 MuiTypography-subtitle2-172 MuiTypography-colorTextSecondary-186"
                />
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root-57 makeStyles-lpLinks-9 MuiGrid-container-58 MuiGrid-item-59"
          >
            <div
              class="MuiGrid-root-57 MuiGrid-container-58 MuiGrid-spacing-xs-2-81 MuiGrid-direction-xs-column-61"
            >
              <div
                class="MuiGrid-root-57 MuiGrid-item-59"
              >
                <p
                  class="MuiTypography-root-160 makeStyles-buyText-228 MuiTypography-body1-162 MuiTypography-colorTextSecondary-186"
                >
                  Buy Badger:
                </p>
                <div
                  class="MuiGrid-root-57 MuiGrid-container-58 MuiGrid-spacing-xs-3-82"
                >
                  <div
                    class="MuiGrid-root-57 MuiGrid-item-59"
                  >
                    <a
                      class="MuiTypography-root-160 MuiLink-root-231 MuiLink-underlineHover-233 makeStyles-link-229 MuiTypography-colorPrimary-183"
                      href="https://curve.fi/factory-crypto/4"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span
                        class="MuiTypography-root-160 makeStyles-linkText-230"
                      >
                        Curve
                      </span>
                      <img
                        alt="Curve link"
                        src="/assets/icons/link-icon.svg"
                      />
                    </a>
                  </div>
                  <div
                    class="MuiGrid-root-57 MuiGrid-item-59"
                  >
                    <a
                      class="MuiTypography-root-160 MuiLink-root-231 MuiLink-underlineHover-233 makeStyles-link-229 MuiTypography-colorPrimary-183"
                      href="https://app.sushi.com/swap?inputCurrency=0x2260fac5e5542a773aa44fbcfedf7c193bc2c599&outputCurrency=0x3472A5A71965499acd81997a54BBA8D852C6E53d"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span
                        class="MuiTypography-root-160 makeStyles-linkText-230"
                      >
                        Sushiswap
                      </span>
                      <img
                        alt="Sushiswap link"
                        src="/assets/icons/link-icon.svg"
                      />
                    </a>
                  </div>
                  <div
                    class="MuiGrid-root-57 MuiGrid-item-59"
                  >
                    <a
                      class="MuiTypography-root-160 MuiLink-root-231 MuiLink-underlineHover-233 makeStyles-link-229 MuiTypography-colorPrimary-183"
                      href="https://app.uniswap.org/#/swap?inputCurrency=0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599&outputCurrency=0x3472A5A71965499acd81997a54BBA8D852C6E53d"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span
                        class="MuiTypography-root-160 makeStyles-linkText-230"
                      >
                        Uniswap
                      </span>
                      <img
                        alt="Uniswap link"
                        src="/assets/icons/link-icon.svg"
                      />
                    </a>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root-57 MuiGrid-item-59"
              >
                <p
                  class="MuiTypography-root-160 makeStyles-buyText-228 MuiTypography-body1-162 MuiTypography-colorTextSecondary-186"
                >
                  Buy Digg:
                </p>
                <div
                  class="MuiGrid-root-57 MuiGrid-container-58 MuiGrid-spacing-xs-3-82"
                >
                  <div
                    class="MuiGrid-root-57 MuiGrid-item-59"
                  >
                    <a
                      class="MuiTypography-root-160 MuiLink-root-231 MuiLink-underlineHover-233 makeStyles-link-229 MuiTypography-colorPrimary-183"
                      href="https://app.sushi.com/swap?inputCurrency=0x2260fac5e5542a773aa44fbcfedf7c193bc2c599&outputCurrency=0x3472A5A71965499acd81997a54BBA8D852C6E53d"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span
                        class="MuiTypography-root-160 makeStyles-linkText-230"
                      >
                        Sushiswap
                      </span>
                      <img
                        alt="Sushiswap link"
                        src="/assets/icons/link-icon.svg"
                      />
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root-57 MuiGrid-item-59"
        >
          <button
            aria-label="disconnect wallet"
            class="MuiButtonBase-root-199 MuiButton-root-237 MuiButton-text-239"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-238 makeStyles-disconnectWalletText-5"
            >
              <img
                alt="disconnect wallet icon"
                aria-label="disconnect wallet"
                class="makeStyles-disconnectWalletIcon-13"
                src="/assets/icons/disconnect-wallet.svg"
              />
              Disconnect Wallet
            </span>
            <span
              class="MuiTouchRipple-root-266"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
